File: dsa-problems/Recursion/solutions/java/REC-008-alternating-vowel-consonant-ladder.java
import java.util.*;

class Solution {
    Set<String> results;
    boolean[] used;
    String S;
    int N;

    public List<String> getAlternatingPermutations(String s) {
        S = s;
        N = s.length();
        results = new TreeSet<>(); // TreeSet for sorted order
        used = new boolean[N];
        backtrack(new StringBuilder());
        return new ArrayList<>(results);
    }

    private boolean isVowel(char c) {
        return "aeiou".indexOf(c) != -1;
    }

    private void backtrack(StringBuilder current) {
        if (current.length() == N) {
            results.add(current.toString());
            return;
        }

        char lastChar = current.length() > 0 ? current.charAt(current.length() - 1) : '\0';
        boolean lastIsVowel = current.length() > 0 ? isVowel(lastChar) : false;

        for (int i = 0; i < N; i++) {
            if (!used[i]) {
                char nextChar = S.charAt(i);
                boolean nextIsVowel = isVowel(nextChar);
                
                if (current.length() == 0 || lastIsVowel != nextIsVowel) {
                    used[i] = true;
                    current.append(nextChar);
                    backtrack(current);
                    current.deleteCharAt(current.length() - 1);
                    used[i] = false;
                }
            }
        }
    }
}

class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if(!sc.hasNext()) return;
        String s = sc.next();
        
        Solution sol = new Solution();
        List<String> res = sol.getAlternatingPermutations(s);
        if(res.isEmpty()) {
            System.out.println("NONE");
        } else {
            for(String p : res) System.out.println(p);
        }
        sc.close();
    }
}

File: dsa-problems/Recursion/solutions/cpp/REC-008-alternating-vowel-consonant-ladder.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <set>

using namespace std;

class Solution {
    string S;
    int N;
    set<string> results;
    vector<bool> used;

public:
    vector<string> getAlternatingPermutations(string s) {
        S = s;
        N = s.length();
        results.clear();
        used.assign(N, false);
        string current = "";
        backtrack(current);
        return vector<string>(results.begin(), results.end());
    }

    bool is_vowel(char c) {
        return string("aeiou").find(c) != string::npos;
    }

    void backtrack(string& current) {
        if (current.length() == N) {
            results.insert(current);
            return;
        }

        for (int i = 0; i < N; i++) {
            if (!used[i]) {
                if (current.empty() || is_vowel(current.back()) != is_vowel(S[i])) {
                    used[i] = true;
                    current.push_back(S[i]);
                    backtrack(current);
                    current.pop_back();
                    used[i] = false;
                }
            }
        }
    }
};

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr);
    string s;
    if (!(cin >> s)) return 0;
    
    Solution sol;
    vector<string> res = sol.getAlternatingPermutations(s);
    if(res.empty()) {
        cout << "NONE" << endl;
    } else {
        for(const string& p : res) cout << p << endl;
    }
    return 0;
}
